### PR TITLE
add staking contracts

### PR DESCRIPTION
**Pull request by coolhill**
**sQuartz** - The yield strategies takes a performance fee which is distributed to Sandclock Quartz holders thru a staking contract deployed on mainnet. These contracts additionally distributes bonus multiplier points (inspired by [GMX reward mechanism](https://gmxio.gitbook.io/gmx/rewards)) in order to incentivize long term staking. Deposit Quartz and receive sQuartz at a 1:1 ratio. Over time user earns a bonus multiplier at 100% APY which can be compounded in order to claim a bigger portion of the staking rewards. Claim staking rewards in WETH over time as performance fees are collected. The amount of staked Quartz and compounded multiplier points among stakers decide reward distribution. Redeem sQuartz for Quartz at any time, which will additionally burn some of the earned multiplier points for that account.

- deployed https://sepolia.etherscan.io/address/0xd5e300dbf4b7b5f4586600395eeca12ab14477c1 for testing

### CODE REVIEW

File: src/staking/BonusTracker.sol [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        uint256 bonusPerToken_ = _bonusPerToken(lastTimeBonusApplicable_);
+
+        // accrue bonus
+        bonusPerTokenStored = bonusPerToken_;
+        lastBonusUpdateTime = lastTimeBonusApplicable_;
+        bonus[msg.sender] = _earnedBonus(msg.sender, accountBalance, bonusPerToken_, bonus[msg.sender]);
+        userBonusPerTokenPaid[msg.sender] = bonusPerToken_;
+    }
+
+    function _updateReward(address) internal virtual {}
```
[the remark this answers is not in the document]
Comment (coolhill) [reply src/staking/BonusTracker.sol:91]: false its used [here](https://github.com/lindy-labs/sandclock-contracts/pull/76/files#diff-b821b5c723eb7bfa158fbd338fa96c47df52547d99731a3a58875c305cd2da98R38)

File: src/staking/BonusTracker.sol [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+import {ERC20} from "solmate/tokens/ERC20.sol";
+import {ERC4626} from "solmate/mixins/ERC4626.sol";
+import {FixedPointMathLib} from "solmate/utils/FixedPointMathLib.sol";
+import {ReentrancyGuard} from "openzeppelin-contracts/security/ReentrancyGuard.sol";
+
+abstract contract BonusTracker is ERC4626, ReentrancyGuard {
+    using FixedPointMathLib for uint256;
+
+    event BonusPaid(address indexed user, uint256 bonus);
```
Comment (coolhill) [inline src/staking/BonusTracker.sol:12]:
```suggestion
    event BonusPaid(address indexed user, uint256 bonus);
    event BonusBurned(address indexed user, uint256 bonus);
```

File: src/staking/RewardTracker.sol [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+            // if they are sending/redeeming everything then burn everything
+            if (amount == balance) {
+                multiplierPointsOf[sender] = 0;
+                totalBonus -= bonus_;
+            } else {
+                // otherwise burn an equivalent percentage
+                uint256 burnAmount = bonus_.mulDivDown(amount, balance);
+                multiplierPointsOf[sender] -= burnAmount;
+                totalBonus -= burnAmount;
+            }
```
Comment (coolhill) [inline src/staking/RewardTracker.sol:233]:
```suggestion
                bonus_ = bonus_.mulDivDown(amount, balance);
                multiplierPointsOf[sender] -= bonus_;
                totalBonus -= bonus_;
            }
            emit BonusBurned(sender, bonus_);
```

File: src/staking/RewardTracker.sol [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        assets = totalSupply;
+    }
+
+    function deposit(uint256 assets, address receiver) public override returns (uint256 shares) {
+        _updateReward(receiver);
+        _updateBonus(receiver);
+        shares = super.deposit(assets, receiver);
+    }
+
+    function mint(uint256 assets, address receiver) public override returns (uint256 shares) {
```
Comment (fyang1024) [inline src/staking/RewardTracker.sol:65]: Why does the `mint` function have the same implementation as the `deposit` function?

File: src/staking/RewardTracker.sol [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        _updateBonus(msg.sender);
+        _updateBonus(to);
+        _burnMultiplierPoints(amount, msg.sender);
+        return super.transfer(to, amount);
+    }
+
+    function transferFrom(address from, address to, uint256 amount) public override returns (bool) {
+        _updateReward(from);
+        _updateReward(to);
+        _updateBonus(msg.sender);
```
Comment (fyang1024) [inline src/staking/RewardTracker.sol:210]:
```suggestion
        _updateBonus(from);
```

File: src/staking/RewardTracker.sol [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    function deposit(uint256 assets, address receiver) public override returns (uint256 shares) {
+        _updateReward(receiver);
+        _updateBonus(receiver);
+        shares = super.deposit(assets, receiver);
+    }
+
+    function mint(uint256 assets, address receiver) public override returns (uint256 shares) {
+        _updateReward(receiver);
+        _updateBonus(receiver);
+        shares = super.deposit(assets, receiver);
```
Comment (coolhill) [inline src/staking/RewardTracker.sol:68]:
```suggestion
    function mint(uint256 shares, address receiver) public override returns (uint256 assets) {
        _updateReward(receiver);
        _updateBonus(receiver);
        shares = super.mint(shares, receiver);
```

File: src/staking/RewardTracker.sol [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    function deposit(uint256 assets, address receiver) public override returns (uint256 shares) {
+        _updateReward(receiver);
+        _updateBonus(receiver);
+        shares = super.deposit(assets, receiver);
+    }
+
+    function mint(uint256 shares, address receiver) public override returns (uint256 assets) {
+        _updateReward(receiver);
+        _updateBonus(receiver);
+        shares = super.mint(shares, receiver);
```
Comment (coolhill) [inline src/staking/RewardTracker.sol:68]:
```suggestion
        assets = super.mint(shares, receiver);
```

File: src/staking/RewardTracker.sol (hunk 4 above)
Comment (coolhill) [reply src/staking/RewardTracker.sol]: fixed :roller_skate:

File: src/staking/RewardTracker.sol [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+        lastUpdateTime = uint64(block.timestamp);
+        periodFinish = uint64(block.timestamp + DURATION_);
+
+        emit RewardAdded(reward);
+    }
+
+    /// @notice Lets a reward distributor fetch performance fees from
+    /// a vault and start a new reward period.
+    function fetchRewards(ERC4626 vault) external onlyRole(DISTRIBUTOR) {
+        require(isVault[address(vault)], "vault not whitelisted");
```
Comment (brkomir) [inline src/staking/RewardTracker.sol:176]: you can introduce a specific error type and use `revert`

File: src/staking/RewardTracker.sol [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+        uint256 beforeBalance = rewardToken.balanceOf(address(this));
+        vault.redeem(vault.balanceOf(address(this)), address(this), address(this));
+        uint256 afterBalance = rewardToken.balanceOf(address(this));
+        _notifyRewardAmount(afterBalance - beforeBalance);
+    }
+
+    /// @notice Lets an admin add a vault for collecting fees from.
+    function addVault(address vault) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(ERC4626(vault).asset() == rewardToken, "only WETH assets");
```
Comment (brkomir) [inline src/staking/RewardTracker.sol:185]: same as above

File: src/staking/RewardTracker.sol [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+        return rewardPerTokenStored
+            + rewardRate_.mulDivDown((lastTimeRewardApplicable_ - lastUpdateTime) * PRECISION, totalSupply_);
+    }
+
+    function _updateReward(address account) internal override {
+        // storage loads
+        uint256 accountBalance = balanceOf[account] + multiplierPointsOf[account];
+        uint64 lastTimeRewardApplicable_ = lastTimeRewardApplicable();
+        uint256 totalSupply_ = totalSupply + totalBonus;
+        uint256 rewardPerToken_ = _rewardPerToken(totalSupply_, lastTimeRewardApplicable_, rewardRate);
```
Comment (brkomir) [inline src/staking/RewardTracker.sol:215]:
```suggestion
        uint256 rewardPerToken_ = _calcRewardPerToken(totalSupply + totalBonus, lastTimeRewardApplicable_, rewardRate);
```

File: src/staking/RewardTracker.sol [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        _burnMultiplierPoints(amount, from);
+        return super.transferFrom(from, to, amount);
+    }
+
+    // burn multiplier points
+    function _burnMultiplierPoints(uint256 amount, address sender) internal {
+        uint256 bonus_ = multiplierPointsOf[sender];
+
+        // if the sender has bonus points
+        if (bonus_ > 0) {
```
Comment (brkomir) [inline src/staking/RewardTracker.sol:247]: just a suggestion to change to `if(bonus == 0) return;` but that's nitpicking

**Review by brkomir — COMMENTED** [review]
left come comments but overall looks good!

File: src/staking/RewardTracker.sol [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    /// @notice Lets a reward distributor fetch performance fees from
+    /// a vault and start a new reward period.
+    function fetchRewards(ERC4626 vault) external onlyRole(DISTRIBUTOR) {
+        require(isVault[address(vault)], "vault not whitelisted");
+        uint256 beforeBalance = rewardToken.balanceOf(address(this));
+        vault.redeem(vault.balanceOf(address(this)), address(this), address(this));
+        uint256 afterBalance = rewardToken.balanceOf(address(this));
+        _notifyRewardAmount(afterBalance - beforeBalance);
```
[the remark this answers is not in the document]
Comment (coolhill) [reply src/staking/RewardTracker.sol]: i don't trust the return value on purpose as a security measure. may be overkill tho

[the remark this answers is not in the document]
Comment (brkomir) [reply src/staking/RewardTracker.sol]: point taken. however, we are the authors of the vault so the return value can be trusted but I don't mind leaving this as is, it's your call

File: src/staking/RewardTracker.sol (hunk 11 above)
Comment (coolhill) [reply src/staking/RewardTracker.sol]: fixed

File: src/staking/RewardTracker.sol (hunk 8 above)
Comment (coolhill) [reply src/staking/RewardTracker.sol]: fixed

File: src/staking/RewardTracker.sol (hunk 9 above)
Comment (coolhill) [reply src/staking/RewardTracker.sol]: fixed

File: src/staking/RewardTracker.sol [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+        if (block.timestamp >= periodFinish_) {
+            newRewardRate = reward / DURATION_;
+        } else {
+            uint256 remaining = periodFinish_ - block.timestamp;
+            uint256 leftover = remaining * rewardRate_;
+            newRewardRate = (reward + leftover) / DURATION_;
+        }
+        // prevent overflow when computing rewardPerToken
+        if (newRewardRate >= ((type(uint256).max / PRECISION) / DURATION_)) {
+            revert Error_AmountTooLarge();
```
[the remark this answers is not in the document]
Comment (coolhill) [reply src/staking/RewardTracker.sol]: fixed

**Review by brkomir — APPROVED** [review]

**Pull request merged**